Add nothrow to visitors (only tried on Windows so far)

n8sh opened the pull request:
I took a stab at converting everything to `nothrow` at once (doing it all at once [is seemingly necessary due to circular dependencies](https://github.com/dlang/dmd/pull/15521#issuecomment-1673650324)). Only tested on my Windows machine so far.

---

To make reviewing this tractable here are the changes that _aren't_ just annotating function definitions with nothrow:

1\. dmd.vsoptions.detectVSInstallDirViaCOM:
\- Cheat by using `scope(failure) assert(0)`. This might defeat the purpose of making things nothrow. This is the sole place we had to resort to that.

2\. dmd.vsoptions.VSOptions.isWin64Host:
\- Assume external function IsWow64Process is nothrow.

3\. dmd.gluelayer
\- (GCC only) Assume external functions asmSemantic and toObjFile are
nothrow.

4\. In these functions replaced lazy parameters with delegate parameters because nothrow is not inferred for lazy arguments (https://issues.dlang.org/show_bug.cgi?id=12647):
\- dmd.cppmangle.Context.push
\- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(ClassDeclaration).resolveBase
\- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(InterfaceDeclaration).resolveBase

The places those functions were called that were lambda-fied:
\- dmd.cppmangle.CppMangleVisitor.headOfType
\- dmd.cppmangle.CppMangleVisitor.mangleFunctionParameters
\- dmd.cppmangle.CppMangleVisitor.mangleReturnType
\- dmd.cppmangle.CppMangleVisitor.template_arg
\- dmd.cppmangle.CppMangleVisitor.writeQualified
\- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(ClassDeclaration)
\- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(InterfaceDeclaration)

5\. In dmd.dtoh.ToCppBuffer.ensureDeclared replaced AA `get` with `in` because `get` doesn't infer nothrow.

6\. Redefined delegate/function pointer types (aliases) to be nothrow:
\- dmd.astbase.ASTBase.Parameter.ForeachDg
\- dmd.errors.DiagnosticHandler
\- dmd.errors.FatalErrorHandler
\- dmd.globals.Global.FileName

7\. Redefined delegate/function pointer fields to be nothrow:
\- dmd.ob.genKill.foreachExp.ExpWalker.dgWriteVar
\- dmd.ob.genKill.foreachExp.ExpWalker.dgReadVar
\- dmd.ob.checkObErrors.foreachExp.ExpWalker.dgWriteVar
\- dmd.ob.checkObErrors.foreachExp.ExpWalker.dgReadVar

8\. Change delegate/function pointer arguments to be nothrow in:
\- dmd.aggregate.apply
\- dmd.argtypes_sysv_x64.ToClassesVisitor.classifyFields
\- dmd.argtypes_x86.toArgTypes_x86.aggregate
\- dmd.cppmangle.CppMangleVisitor.writeChained
\- dmd.cppmangle.CppMangleVisitor.writeQualified
\- dmd.dsymbol.foreachDsymbol
\- dmd.foreachvar.foreachVar . dmd.foreachvar.foreachVar.VarWalker.this
\- dmd.initsem.resolveStructLiteralNamedArgs
\- dmd.mtype.modifiersApply
\- dmd.mtype.attributesApply
\- dmd.ob.genKill.foreachExp.ExpWalker.this
\- dmd.ob.checkObErrors.foreachExp.ExpWalker.this
\- dmd.objc_glue.Symbols.cache
\- dmd.scanmscoff.scanMSCoffObjModule
\- dmd.scanomf.scanOmfObjModule
\- dmd.scanomf.scanOmfLib
\- dmd.statementsem.statementSemanticVisit.visitReturn.turnOffRef

n8sh commented:
[Weird build error.](https://dtest.dlang.io/results/b52acc05a6d3340e55efd61164a928bad0edb7d5/6b54c3f2fcd39bfcd6c3fc4bce82e8f43ac11b77/build.log)

```console
src/dmd/glue.d(497): Error: function `core.stdc.stdlib.qsort` is not `nothrow`
src/dmd/glue.d(429): Error: `nothrow` function `dmd.glue.genObjFile` may throw
```

Yet:
https://github.com/dlang/dmd/blob/5fb9de2bfb45f5e374d4990d588663ec6c0c936f/druntime/src/core/stdc/stdlib.d#L40-L46

dkorpel commented:
Some CI are building DMD with an older version (2.079) for bootstrapping, where certain annotations in druntime are missing. A solution is to define these extern functions with attributes yourself instead of importing them.

maxhaton commented:
the array module declares it's own qsort, there's precedent for it.

maxhaton commented:
I don't really see what the point is in spending much time on this though. For performance? The only thing my CPU and I care about is the amount of memory allocated i.e. both in the size of types and the volume of allocations.

Actually thinking about it this could break LDC because they do use Phobos in their diff to the frontend, which includes poking around in the sema.

ibuclaw commented on compiler/src/dmd/attrib.d:
```diff
@@ -520,22 +520,23 @@ extern (C++) final class CPPMangleDeclaration : AttribDeclaration
 extern (C++) final class CPPNamespaceDeclaration : AttribDeclaration
 {
+nothrow:
     /// CTFE-able expression, resolving to `TupleExp` or `StringExp`
     Expression exp;
 
-    extern (D) this(const ref Loc loc, Identifier ident, Dsymbols* decl) @safe
+    extern (D) this(const ref Loc loc, Identifier ident, Dsymbols* decl) nothrow @safe
```
I'd have thought that most the time all you need is `nothrow:` at the top of the class/struct member list, and not adding `nothrow` to every single member individually.

n8sh replied:
Yes. If that way is considered better style it certainly makes a smaller diff.

n8sh replied:
Changed to do things this way within class/struct declarations but not at the tops of modules.

ibuclaw commented on compiler/src/dmd/cppmangle.d:
```diff
@@ -137,18 +137,27 @@ private struct Context
      */
-    private Context push(lazy RootObject next) @safe
+    private Context push(RootObject delegate() nothrow @safe next) nothrow @safe
     {
         auto r = this.res;
         if (r !is null)
-            this.res = next;
+            this.res = next();
+        return Context(this.ti, this.fd, r);
+    }
+
+    // ditto
+    private Context push(RootObject delegate() nothrow @system next) nothrow @system
+    {
+        auto r = this.res;
+        if (r !is null)
+            this.res = next();
```
This kind of change is one that can be broken out into a separate PR - [refactor] Replace all lazy parameters with `@safe` and `@system` delegates.

ibuclaw commented on compiler/src/dmd/dtoh.d:
```diff
@@ -2921,8 +2921,18 @@ public:
 
         // Omit redundant fwd-declaration if we already emitted the entire declaration
-        if (visited.get(check, false))
-            return;
+        version (none)
+        {
+            // FIXME: use this version when 'get' can infer nothrow
+            if (visited.get(check, false))
+                return;
+        }
+        else
+        {
+            if (const pVisited = check in visited)
+                if (*pVisited)
+                    return;
+        }
```
Note, this code should be compilable by compilers as old as 2.076. If `nothrow` isn't inferred, then you'll have to make it explicit.

n8sh replied:
[Per run.dlang.io the following code doesn't compile with any version of DMD:](https://run.dlang.io/is/5DfcGK)
```d
int main() nothrow
{
    int[int] aa;
    return aa.get(1, 0);
}
```

ibuclaw commented on compiler/src/dmd/iasm.d:
```diff
@@ -35,7 +35,7 @@ else
 /************************ AsmStatement ***************************************/
 
-extern(C++) Statement asmSemantic(AsmStatement s, Scope *sc)
+extern(C++) Statement asmSemantic(AsmStatement s, Scope *sc) nothrow
```
You'll need to compile in `iasmgcc.d` instead of `iasmdmd.d`, as I'm sure there's be more nothrow violations in that module (used by GDC and LDC).

n8sh committed:
Add nothrow to visitors

To make reviewing this tractable here are the changes that _aren't_ just
annotating function definitions with nothrow:

1. dmd.vsoptions.detectVSInstallDirViaCOM:
- Cheat by using `scope(failure) assert(0)`. This might defeat the
purpose of making things nothrow. This is the sole place we had to
resort to that.

2. dmd.vsoptions.VSOptions.isWin64Host:
- Assume external function IsWow64Process is nothrow.

3. dmd.gluelayer
- (GCC only) Assume external functions asmSemantic and toObjFile are
nothrow.

4. In these functions replaced lazy parameters with delegate parameters
   because nothrow is not inferred for lazy arguments (https://issues.dlang.org/show_bug.cgi?id=12647):
- dmd.cppmangle.Context.push
- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(ClassDeclaration).resolveBase
- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(InterfaceDeclaration).resolveBase

The places those functions were called that were lambda-fied:
- dmd.cppmangle.CppMangleVisitor.headOfType
- dmd.cppmangle.CppMangleVisitor.mangleFunctionParameters
- dmd.cppmangle.CppMangleVisitor.mangleReturnType
- dmd.cppmangle.CppMangleVisitor.template_arg
- dmd.cppmangle.CppMangleVisitor.writeQualified
- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(ClassDeclaration)
- dmd.dsymbolsem.DsymbolSemanticVisitor.visit(InterfaceDeclaration)

5. In dmd.dtoh.ToCppBuffer.ensureDeclared replaced AA `get` with `in`
because `get` doesn't infer nothrow.

6. Added nothrow definition of qsort to dmd.glue

7. Redefined delegate/function pointer types (aliases) to be nothrow:
- dmd.astbase.ASTBase.Parameter.ForeachDg
- dmd.errors.DiagnosticHandler
- dmd.errors.FatalErrorHandler
- dmd.globals.Global.FileName

8. Redefined delegate/function pointer fields to be nothrow:
- dmd.ob.genKill.foreachExp.ExpWalker.dgWriteVar
- dmd.ob.genKill.foreachExp.ExpWalker.dgReadVar
- dmd.ob.checkObErrors.foreachExp.ExpWalker.dgWriteVar
- dmd.ob.checkObErrors.foreachExp.ExpWalker.dgReadVar

9. Change delegate/function pointer arguments to be nothrow in:
- dmd.aggregate.apply
- dmd.argtypes_sysv_x64.ToClassesVisitor.classifyFields
- dmd.argtypes_x86.toArgTypes_x86.aggregate
- dmd.cppmangle.CppMangleVisitor.writeChained
- dmd.cppmangle.CppMangleVisitor.writeQualified
- dmd.dsymbol.foreachDsymbol
- dmd.foreachvar.foreachVar
. dmd.foreachvar.foreachVar.VarWalker.this
- dmd.initsem.resolveStructLiteralNamedArgs
- dmd.mtype.modifiersApply
- dmd.mtype.attributesApply
- dmd.ob.genKill.foreachExp.ExpWalker.this
- dmd.ob.checkObErrors.foreachExp.ExpWalker.this
- dmd.objc_glue.Symbols.cache
- dmd.scanmscoff.scanMSCoffObjModule
- dmd.scanomf.scanOmfObjModule
- dmd.scanomf.scanOmfLib
- dmd.statementsem.statementSemanticVisit.visitReturn.turnOffRef

n8sh commented:
```console
src/dmd/cppmangle.d(1364): Error: cannot implicitly convert expression __lambda3 of type RootObject delegate() nothrow to RootObject delegate() nothrow @safe
```
```d
            auto prev = this.context.push(() {
                    TypeFunction tf;
                    if (isDsymbol(this.context.res))
                        tf = this.context.res.asFuncDecl().type.isTypeFunction();
                    else
                        tf = this.context.res.asType().isTypeFunction();
                    assert(tf);
                    return (*tf.parameterList.parameters)[n].type;
                });
```
Now this is odd since there are versions for both safe and unsafe delegates.
```d
    private Context push(RootObject delegate() nothrow @safe next) @safe { ... }
    private Context push(RootObject delegate() nothrow @system next) @system { ... }
```
I'll see if changing `push` to a template stops (some) compilers from complaining about that.

maxhaton commented:
@kinke @JohanEngelen see comment passim about ldc

JohanEngelen commented:
If this means one cannot use (many of the) Phobos functions inside a visitor, then that is pretty bad yes. Think about dmd as library use.
What's the benefit of this change?

WalterBright commented:
With the new `-nothrow` switch this is no longer necessary.
